src/goto-instrument/function_stubs.cpp

```diff
@@ -0,0 +1,264 @@
+  const symbolt &f_sym = ns.lookup(function_id);
+  const code_typet &type = to_code_type(f_sym.type);
+
+  const exprt &requires =
+    static_cast<const exprt &>(type.find(ID_C_spec_requires));
+  const exprt &ensures =
+    static_cast<const exprt &>(type.find(ID_C_spec_ensures));
+
+  // Clear the function body
+  goto_function.body.clear();
+
+  // Add the assertion in the function body
+  if(requires.is_not_nil())
+    goto_function.body.add(
+      goto_programt::make_assertion(requires, requires.source_location()));
+
+  // Get the implementation function symbol
+  //
+  // Question: Should I do lookups on the symbol table or should I do
+  // them on the namespace?
+  const symbolt &impl_f_sym =
+    symbol_table.lookup_ref(stub_name_of_function(function_id));
+
+  // Make the function call to the implementation function
+  code_function_callt call(impl_f_sym.symbol_expr());
+
+  // Declare a variable to store the return value from the internal call
+  symbol_exprt r = get_fresh_aux_symbol(
+                     type.return_type(),
+                     id2string(function_id) + "::ret_val",
+                     "ret_val",
+                     source_locationt(),
+                     impl_f_sym.mode,
+                     symbol_table)
+                     .symbol_expr();
+
+  replace_symbolt replace;
+
+  if(type.return_type() != empty_typet())
+  {
+    // Declare the return variable and assign to it the return value
+    // of the call
+    goto_function.body.add(goto_programt::make_decl(r));
+    call.lhs() = r;
+
+    // Replace the return_value with r
+    symbol_exprt ret_val(CPROVER_PREFIX "return_value", type.return_type());
+    replace.insert(ret_val, r);
+  }
+
+  // Add the parameters as arguments to the call
+  auto parameter_identifiers = type.parameter_identifiers();
+  for(auto it = parameter_identifiers.begin();
+      it != parameter_identifiers.end();
+      it++)
+  {
+    const symbolt &parameter_symbol = ns.lookup(*it);
+    call.arguments().push_back(parameter_symbol.symbol_expr());
+  }
+
+  goto_function.body.add(
+    goto_programt::make_function_call(code_function_callt(call)));
+  // Maybe: Add location/comments maybe?
+
+  // Add the postcondition assumption
+  if(ensures.is_not_nil())
+  {
+    exprt ensures_replaced = ensures;
+    replace(ensures_replaced);
+    goto_function.body.add(goto_programt::make_assumption(
+      ensures_replaced, ensures.source_location()));
+  }
+
+  if(type.return_type() != empty_typet())
+  {
+    code_returnt return_expr(r);
+    goto_function.body.add(goto_programt::make_return(return_expr));
+  }
+
+  // Append the end of function instruction
+  goto_function.body.add(
+    goto_programt::make_end_function(ensures.source_location()));
+
+  log.debug() << "Function body after adding the assumption: "
+              << "\n";
+  goto_function.body.output(ns, function_id, log.debug());
+
+  // Question: do I need to call some update or something after I do
```

Comment:
Yep, `goto_model.update()` after the whole lot is probably right

Reply:
thanks